File: Go/math/sieve_of_eratosthenes.go
package main

import "fmt"

func Sieve(n int) {

	prime := make([]bool, n+1)

	for itr := 0; itr <= n; itr++ {
		prime[itr] = true
	}

	for p := 2; p*p <= n; p++ {
		if prime[p] == true {
			for x := p * p; x <= n; x += p {
				prime[x] = false
			}
		}
	}

	for p := 2; p <= n; p++ {
		if prime[p] == true {
			fmt.Print(p, " ")
		}
	}
}

func main() {
	var num int

	fmt.Print("Enter the range: ")
	fmt.Scan(&num)
	Sieve(num)
}

// TIME COMPLEXITY: O(n*log(log(n)))
// SPACE COMPLEXITY: O(n)

// TEST CASE 1:
// Enter the range: 50
// 2 3 5 7 11 13 17 19 23 29 31 37 41 43 47

// TEST CASE 2:
// Enter the range: 100
// 2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97

File: Python/math/Sieve-of-eratosthenes.py
"""
The sieve of Eratosthenes is an algorithm for finding all prime numbers up to any given limit. 
It is computationally highly efficient algorithm.Let us take the number `n = 100` , so we create list from `2` to `100` . According to the algorithm , we mark all the numbers which are divisible by `2` which are greater than or equal to the square of it . Then we move to the next number `3` , 
we mark all the numbers which are divisible by `3`  which are greater than or equal to the square of it. 
we continue this process until the square of the number is greater than `n`. 
The numbers which are unmarked in the list are the prime numbers less than `n`.
"""


def sieve_of_eratosthenes(n):
    sieve = [True for i in range(n + 1)]
    p = 2
    while p ** 2 <= n:
        if sieve[p] == True:
            i = p * p
            while i <= n:
                sieve[i] = False
                i += p
        p += 1
    for i in range(2, len(sieve)):
        if sieve[i] == True:
            print(i, end=" ")


def main():
    print("Enter the number upto which prime numbers are to be computed: ")
    n = int(input())
    print("The number of prime numbers less than " + str(n) + " is :")
    sieve_of_eratosthenes(n)


if __name__ == "__main__":
    main()


"""
Sample I/O:

Enter the number upto which prime numbers are to be computed:
30
The number of prime numbers less than 30 is :
2 3 5 7 11 13 17 19 23 29

Time complexity : O(n*(log(log(n))))
Space complexity : O(n)
"""
